File: src/tmp/Python/array/contiguous_array.py
from typing import List

def findMaxLength(nums: List[int]) -> int:
    sum_0, sum_1, result, remain = 0, 0, 0, {}
    for i in range(len(nums)):
        if nums[i] == 0:
            sum_0 += 1
        else:
            sum_1 += 1
        
        num = sum_0 - sum_1
        if num == 0:
            result = max(result, i + 1)
        elif remain.get(num, 50000) != 50000:
            result = max(result, i - remain[num])
        else:
            remain[num] = i
    
    return result

if __name__ == "__main__":
    print(findMaxLength([0,1]))

File: core/tmp/Java/src/com/dbc/algorithm/Array/ContiguousArray.java
package com.dbc.algorithm.Array;

import java.util.HashMap;
import java.util.Map;

public class ContiguousArray {
    public static int findMaxLength(int[] nums) {
        int sum0 = 0, sum1 = 0, result = 0;
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == 0) sum0++;
            else sum1++;

            int num = sum0 - sum1;
            if (num == 0) {
                result = Math.max(result, i + 1);
            } else if (map.getOrDefault(num, 50000) != 50000) {
                result = Math.max(result, i - map.get(num));
            } else {
                map.put(num, i);
            }
        }

        return result;
    }
}
